Add support for conversion to NWB using ndx beadl

oruebel opened the pull request:
This PR refactors the convert_cli script to support usage of the ndx-beadl extension for storing ARControl data

oruebel committed:
Update gitignore

oruebel committed:
Update the convert_cli script to add basic support for ndx-beadl

oruebel committed:
Update the convertion example notebook

oruebel committed:
Add missing time offset

oruebel committed:
Fix flake8 for convert script

oruebel committed:
Added state transition probability plot to the example notebook

oruebel committed:
Add note on how to define trials in ARControl

oruebel committed:
Add add_arc_to_nwbfile function to allow adding of arc-data to in-memory file

oruebel committed:
modularize the add_arc_to_nwbfile by placing the different paths in separate functions

oruebel committed:
Update variable names for consistency

oruebel committed:
Update internal helper function names for consistency

oruebel committed:
Sort events, actions, and states by time

oruebel committed:
Add TrialTable for ndx-beadl conversion

oruebel committed:
Fix flake8 issue

oruebel committed:
Fix flake8 issue

oruebel committed:
Minory fixes to docs

oruebel committed:
Updated return value for convert to the filename

oruebel committed:
Add ToDo item

oruebel committed:
Modularize ndx-beadl conversion function and add optional parsing of JSON states

oruebel committed:
Add JSON metadata about states

oruebel committed:
Update ToDo items

oruebel committed:
Add TODO items

oruebel commented:
@chenxinfeng4 the conversion is working. There are a few TODO items marked in the code that would be great if you could have a look at to see if you have any thoughts or comments.

oruebel commented on arcontrol2nwb/converter_cli.py:
```diff
@@ -127,88 +262,633 @@ def c_create(MAT):
+    curr_state_end_index = 0
+    for component_name, start_time, stop_time in sorted(trials, key=lambda t: t[1]):
+        # Find the end index for the actions, events, and states within the trial
+        # TODO Multiple  events/actions/states can have the same timestamp. Check that those are assigned correctly.
```
chenxinfeng4 in the ARControl output there are times when events have the exact same timestamp. Here I'm trying to determine which events belong to which component/trial based on the timing. However, how do I determine whether an event belongs to a given component when the timestamp is the same as start/stop of the event, i.e., is an event that starts at the same time as a component considered to be part of that component or is it the last event of the previous component. Similarly, if an event has the same time as the end of the component, should it be counted as part of the component or as the start of the next one?


https://github.com/oruebel/ArControl-convert2-nwb/blob/063e674e1cf463d475071520c075c98a494541e8/playground/2022-1113-224711.txt#L53-L54

chenxinfeng4 replied:
The original log
```
OUT1:26261 1000
C1S1:26261
C1S2:27261
C1S3:32261
C1S5:34261
%2
OUT2:37681 2000
C2S1:37681
```

Equals  to
```
C1S1:26261
OUT1:26261 1000 #belongs to C1S1
C1S2:27261
...
C2S1:37681
OUT2:37681 2000  #belongs to C2S1
```
The OUTPUT logging is always just one step ahead of the STATE logging for the sake of better close loop stimulation.  In short, if an action(output) starts in the joint of STATE1 END/ STATE2 BEGIN, it belongs to **STATE2**; if a event(input) starts in the joint of STATE1 END/STATE2 BEGIN, it belongs to **STATE1**.

oruebel commented on arcontrol2nwb/converter_cli.py:
```diff
@@ -127,88 +262,633 @@ def c_create(MAT):
         ddt = 0.0001
-        tend[tend==tbg_pre] += ddt
-        tbg =  tbg_pre + ddt
+        tend[tend == tbg_pre] += ddt
+        tbg = tbg_pre + ddt
         tend_post = tend + ddt
         seq_t_1 = np.concatenate((tbg, tend))
         seq_t_0 = np.concatenate((tbg_pre, tend_post))
         seq_v_1 = np.ones(seq_t_1.shape)
         seq_v_0 = np.zeros(seq_t_0.shape)
-        seq_t = np.concatenate((seq_t_0, seq_t_1, [time_len+ddt]))
+        seq_t = np.concatenate((seq_t_0, seq_t_1, [time_len + ddt]))
         seq_v = np.concatenate((seq_v_0, seq_v_1, [0]))
-        if not np.any(seq_t==0.0):
+        if not np.any(seq_t == 0.0):
             seq_t = np.append(seq_t, 0.0)
             seq_v = np.append(seq_v, 0)
```
@chenxinfeng4 could you clarify what is code is doing? In particular, it is unclear to me:

 1) how is ``ddt`` being used here?  It looks like it used as some form of time delta to adjust timestamps. Why do timestamps need to be changed? 
2) It looks like a single event ``e`` with timerange ``v`` is here being transformed into a sequence with multiple steps. Is the purpose of this to put `0` values before and after the actual occurrence of the event which is marked by `1`?

chenxinfeng4 replied:
![image](https://user-images.githubusercontent.com/17546446/220948045-243307d5-2466-4961-b3ad-e579bc111a55.png)

One event is originally defined as [tbg, tend], or [tbg, tdur]. 
However, when converting to NWB timeseries, it should be [tbg_pre, tbg, tend, tend_post].

chenxinfeng4 replied:
Please replace that code by this one, and see where it can give the correct result.
```python
        tbg, tend = v[:, 0], v[:, 0] + v[:, 1]
        ddt = 0.0001
        seq_t, seq_v = [], []

        for tbg_now, tend_now in zip(tbg, tend):
            seq_t.extend([tbg_now, tbg_now + ddt, tend_now, tend_now + ddt])
            seq_v.extend([0, 1, 1, 0])

        # insert (x=0.0, y=0) in the beginning
        if 0.0 not in seq_t:
            seq_t.insert(0, 0.0)
            seq_v.insert(0, 0)
        # insert (x=END, y=0) in the end
        seq_t.append(time_len + ddt)
        seq_v.append(0)
        seq_v = np.array(seq_v)
        seq_t = np.array(seq_t)
        assert np.all(seq_t>=0)
```

chenxinfeng4 replied:
I'll fix that by myself.

oruebel replied:
Thanks for the clarification. That makes sense. For the ndx-beadl representation of the data we then should not need to do these transformation since the timings are here represented via ``start_time, end_time`` (for states) and ``timestep, duration`` (for events/actions), so that this is more closely aligned with how ARControl describes it. The main part then to look out for is that the duration for events/action is being set correctly.

chenxinfeng4 commented:
You've done very nice work. Some features look surprising to me. Some time I don't know what's the meaning of your code. So, can you post your arcontrol the `task json` file (I don't know it) and the `task aconf` file? As well as the breadl `XML/XSD/schema` files, I am not familiar with those.

I have a deadline work to do, but will be back to the issue in 4 days. I'll check the issues you mentioned.

chenxinfeng4 committed:
Add argparse to the CLI.

Add argparse to the CLI. Allow multiple parameters.

chenxinfeng4 committed:
Update converter_cli.py

Fix typo.  Ignore end state event.

chenxinfeng4 committed:
Delete setup.cfg

The setup.cfg is optional.

oruebel commented:
Sorry for the slow reply, I'll get back to this early next week. 

> As well as the beadl `XML/XSD/schema` files, I am not familiar with those.

I'm not converting the ARControl aconf task definition to BEADL, but I'm just saving the aconf file as is in the ``TaskProgram`` field. The ``TaskSchema`` field is for XML schema that is used to define the task program. If ARControl does not formally define a schema for the XML file, then we can leave that field empty. 

> So, can you post your arcontrol the `task json` file (I don't know it) and the `task aconf` file?

I will check with the team that generated the test data I used to make sure I can share the data. I think it should be fine, but want to make sure. The `json` file I think might have been from https://github.com/chenxinfeng4/ArControl-convert2-smcat but I'm not sure. If that is not a common file for ARControl, then it's probably best to remove that part from the script. I was mainly using the JSON to extract some additional annotations for the state definitions. However, that is not required and is probably something we could (with some extra work) also extract from aconf XML file instead.

chenxinfeng4 committed:
Fix import error while beadl not installed

1. Fix import error while beadl not installed.
2. Polish nwb timestamp creation.

chenxinfeng4 commented:
> The json file I think might have been from https://github.com/chenxinfeng4/ArControl-convert2-smcat but I'm not sure

The ArControl didn't use json file yet. You can remove the `arc_taskprogram_json` option.

>  If ARControl does not formally define a schema for the XML file, then we can leave that field empty.

It's OK to leave the `XSD/schema` field empty.
